2022/01/how-to-use-regex-groups-quantifier-in-postgresql.md

```diff
@@ -0,0 +1,86 @@
+
+```bash
+$ content='||121212^^^2^ID 1|676767||SELVA^KUMAR^^^^|19480203|M||B||123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York||123456-7890|||M|NON|4000|'
+
+$ echo $content | grep -Po '([A-Za-z0-9 ]*\^){8}'
+123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^
+```
+
+### PostgreSQL Regex
+
+PostgreSQL's `regexp_matches` function supports the pattern extraction by matching data from the content. While the same perl regex pattern was added to the `regexp_matches` function, instead of all eight groups, only the eighth value was returned. 
+
+```postgres
+=# SELECT REGEXP_MATCHES('||121212^^^2^ID 1|676767||SELVA^KUMAR^^^^|19480203|M||B||123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York||123456-7890|||M|NON|4000|', E'([A-Za-z0-9 ]*\\^){8}[A-Za-z0-9 ]*', 'g');
+ regexp_matches 
+----------------
+ {^}
+(1 row)
+```
+
+More generally, our query is returning the Nth matching group instead of returning all matching groups until the Nth regex group. So if we try to fetch the text matching with 3 groups, the quantifier three will return the third field of all match sections instead of the third group itself.
+
+```postgres
+=# SELECT REGEXP_MATCHES('||121212^^^2^ID 1|676767||SELVA^KUMAR^^^^|19480203|M||B||123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York||123456-7890|||M|NON|4000|', E'([A-Za-z0-9 ]*\\^){3}[A-Za-z0-9 ]*', 'g');
+   regexp_matches   
+--------------------
+ {^}
+ {^}
+ {"New York City^"}
+ {USA^}
+(4 rows)
+```
+
+### PLPERL Function
+
+Since the regexp_matches function doesn’t satisfy our requirements, I attempted to use perl regex through the plperl function. The plperl function also created with the perl regex pattern but the plperl regex function resulted in the same behavior.
+
+```postgres
+=# CREATE OR REPLACE FUNCTION perl_regexp_matches (text) RETURNS text AS $$
+    # PL/Perl function body
+    $input = $_[0];
+    ($output) = $input =~ m/([A-Za-z0-9 ]+ ){2}/g;
+    print $output;
+    return $output;
+$$ LANGUAGE plperl;
+
+=# SELECT perl_regexp_matches('||121212^^^2^ID 1|676767||SELVA^KUMAR^^^^|19480203|M||B||123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York||123456-7890|||M|NON|4000|');
+ perl_regexp_matches 
+---------------------
+ SAMPLE 
+```
+
+### Solution
+
+In regular Postgres expressions, parentheses ( ) create a numbered capture group which leads to returning the quantitative matching results. To get the entire matching data, the regex should have a question mark (?) and a colon (:) added at the beginning of the regex pattern to create a non-capturing group to receive the complete matching group.
+
+```postgres
+=# SELECT REGEXP_MATCHES('||121212^^^2^ID 1|676767||SELVA^KUMAR^^^^|19480203|M||B||123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York||123456-7890|||M|NON|4000|', E'(?:[A-Za-z0-9 ]*\\^){8}[A-Za-z0-9 ]*', 'g');
```

Comment:
Your final solution looks good, though I think it is a bit more complicated than it needs to be.

You can avoid getting the result in an array with only 1 element, and instead just get the string back that you want, using the Postgres `substring` function like this:

```sql
=> SELECT substring('||121212^^^2^ID 1|676767||SELVA^KUMAR^^^^|19480203|M||B||123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York||123456-7890|||M|NON|4000|' FROM '(?:[A-Za-z0-9 ]*\^){8}[A-Za-z0-9 ]*');
                         substring                          
------------------------------------------------------------
 123456 SAMPLE ROAD^^New York City^NY^12345^USA^H^^New York
(1 row)
```

Note also that the `E'...'` quoting just makes life harder since you have to double your `\` escapes. But with standard SQL `'` you don't need to escape `\`.

And you also don't need the `'g'` global match operation, since you're only doing a single match.

Reply:
thanks Jon, your query looks simple and neat. I have updated the final query as you have suggested.